### PR TITLE
feat: add initial role api code

### PR DESCRIPTION
**Pull request by kleyow**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -170,18 +170,42 @@ jobs:
...
+            echo ${NVM_DIR}
+            [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"
+            nvm install v12.16.1
+            nvm alias default v12.16.1
+      - run:
+          name: install modules
+          command: |
+            [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"
+            node --version
+            npm ci
```
Comment (kirgene) [inline .circleci/config.yml:185]: What's the reason for nvm usage? I think we should just use the latest nodejs LTS.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,36 @@
+FROM node:14.3.0-alpine as builder
+USER root
```
Comment (kirgene) [inline Dockerfile:2]: I think it's worth changing the version to `lts-alpine`

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+COPY package.json package-lock.json* /opt/role-assignment-service/
+RUN npm ci
+
+# check in .dockerignore what is skipped during copy
+COPY . .
+
+# cleanup
+RUN apk del build-dependencies
+
+FROM node:14.3.0-alpine
```
Comment (kirgene) [inline Dockerfile:20]: I think it's worth changing the version to `lts-alpine`

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+# Create empty log file & link stdout to the application log file
+RUN mkdir ./logs && touch ./logs/combined.log
+RUN ln -sf /dev/stdout ./logs/combined.log
+
+# Create a non-root user: ra-user
+RUN adduser -D ra-user
+USER ra-user
+
+COPY --chown=ra-user --from=builder /opt/role-assignment-service .
```
Comment (kirgene) [inline Dockerfile:31]: Docker engine does a great job isolating the running process inside a docker container already, so I don't think we need to configure a custom user.

File: docker-compose.yaml [hunk 5]
```diff
@@ -0,0 +1,169 @@
+version: "3.7"
+
+networks:
+  mojaloop-net:
+    name: mojaloop-net
+
+services:
+  mysql-is:
+    container_name: mysql-is
+    image: mysql:5.7.29
```
Comment (kirgene) [inline docker-compose.yaml:10]: I think we should use an up-to-date version of MySQL, i.e. `mysql:8`, as v5 is not actively maintained.

File: src/server/handlers/users/{ID}/roles.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    const userId = request.params.ID
+    const response = await h.getKetoReadApi().getRelationTuples(
+      'role',
+      undefined,
+      'member',
+      userId
+    )
+    const rolesIdList = response.data.relation_tuples?.map(function (relationTuple) {
+      return relationTuple.object
+    })
```
Comment (kirgene) [inline src/server/handlers/users/{ID}/roles.ts]:
```suggestion
    const rolesIdList = response.data.relation_tuples?.map(({object}) => object)
```

File: src/server/handlers/users/{ID}/roles.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+      return {
+        action: operation.action,
+        relation_tuple: {
+          namespace: 'role',
+          object: operation.roleId,
+          relation: 'member',
+          subject: userId
+        }
+      }
+    })
```
Comment (kirgene) [inline src/server/handlers/users/{ID}/roles.ts]:
```suggestion
    const roleDelta: PatchDelta[] = payload.roleOperations.map((operation) => ({
        action: operation.action,
        relation_tuple: {
          namespace: 'role',
          object: operation.roleId,
          relation: 'member',
          subject: userId
        }
    }))
```

File: src/server/handlers/participants.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+  accounts: Record<string, unknown>[];
+}
+
+const get = async (_context: unknown, _request: Request, h: StateResponseToolkit): Promise<ResponseObject> => {
+  try {
+    const response = await axios.get(`${Config.CENTRAL_SERVICE_ADMIN_URL}/participants`)
+    const participantList = response.data as Participant[]
+    const participantIdList = participantList.map(function (obj) {
+      return obj.name
+    })
```
Comment (kirgene) [inline src/server/handlers/participants.ts]:
```suggestion
    const participantIdList = participantList.map((obj) => obj.name)
```

File: src/server/handlers/users.ts [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    })
+    const query = response.data as Wso2IsUserQueryResponse
+    const userIdList = query.Resources.map(function (obj) {
+      return {
+        id: obj.id,
+        name: obj.name,
+        username: obj.userName,
+        emails: obj.emails
+      }
+    })
```
Comment (kirgene) [inline src/server/handlers/users.ts]:
```suggestion
    const userIdList = query.Resources.map((obj) => ({
        id: obj.id,
        name: obj.name,
        username: obj.userName,
        emails: obj.emails
    }))
```

File: src/server/handlers/users/{ID}/participants.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    const userId = request.params.ID
+    const response = await h.getKetoReadApi().getRelationTuples(
+      'participant',
+      undefined,
+      'member',
+      userId
+    )
+    const participantIdList = response.data.relation_tuples?.map(function (relationTuple) {
+      return relationTuple.object
+    })
```
Comment (kirgene) [inline src/server/handlers/users/{ID}/participants.ts]:
```suggestion
    const participantIdList = response.data.relation_tuples?.map(({object}) => object)
```

File: src/server/handlers/users/{ID}/participants.ts [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+      return {
+        action: operation.action,
+        relation_tuple: {
+          namespace: 'participant',
+          object: operation.participantId,
+          relation: 'member',
+          subject: userId
+        }
+      }
+    })
```
Comment (kirgene) [inline src/server/handlers/users/{ID}/participants.ts]:
```suggestion
    const participantDelta: PatchDelta[] = payload.participantOperations.map((operation) => ({
        action: operation.action,
        relation_tuple: {
          namespace: 'participant',
          object: operation.participantId,
          relation: 'member',
          subject: userId
        }
    }))
```

**Review by kirgene — CHANGES_REQUESTED** [review]
Looks good, I put some minor suggestions

File: src/server/handlers/users/{ID}/roles.ts (hunk 6 above)
Comment (kleyow) [reply src/server/handlers/users/{ID}/roles.ts]: this is a result of divergance in eslint config accross projects...and my function autocomplete.

I wonder if there is a proper master config file somewhere so I can stop battling linting...

I've changed to arrow functions here and throughout for other comments

File: docker-compose.yaml (hunk 5 above)
Comment (kleyow) [reply docker-compose.yaml:10]: I pulled this from a wso2 example. https://github.com/wso2/docker-is/blob/master/docker-compose/is/docker-compose.yml
I'd rather not bump it past the example in *this* pr in fear of breaking something, but I *will* try upgrading it in another pr.

File: Dockerfile (hunk 4 above)
Comment (kleyow) [reply Dockerfile:31]: Hmm I'm not quite sure why there was a custom user created but I believe there is an actual reason.
All other microservices follow this dockerfile pattern so I don't want to break away atm.

another standardization thing a template would help resolve over time...

File: Dockerfile (hunk 2 above)
Comment (kleyow) [reply Dockerfile:2]: Lets leave that for the follow up pr.

File: .circleci/config.yml (hunk 1 above)
Comment (kleyow) [reply .circleci/config.yml:185]: Not the initial author but I like nvm for version control. 

> I think we should just use the latest nodejs LTS.

I agree, we are transistioning to the latest LTS iirc... this is a carry-over.
Lets leave that for a follow up ci focused pr.

I really should make a microservice template in mojaloop to stop these carryovers from other services...

**Review by kirgene — APPROVED** [review]

**Pull request merged**